jujugui/static/gui/src/app/state/state.js

```diff
@@ -117,30 +149,192 @@ const State = class State {
   }
 
   /**
+    Stores the dispatchers that are to be called when the appropriate state
+    changes in the application.
+    @param {Array} dispatchers - An array of dispatchers in the format:
+      [['section', callback], ...]
+  */
+  register(dispatchers) {
+    const stored = this._dispatchers;
+    dispatchers.forEach(dispatcher => {
+      const record = [dispatcher[1], dispatcher[2]];
+      if (!stored[dispatcher[0]]) {
+        stored[dispatcher[0]] = [record];
+        return;
+      }
+      stored[dispatcher[0]].push(record);
+    });
+  }
+
+  /**
     Checks the current location and parses it, building the state, then
     executing the registered dispatchers.
+    @param {Array} nullKeys - A list of keys which we must run the 'cleanup'
+      dispaches on before dispatching the current state.
+    @param {Boolean} updateState - Whether this dispatch should update the
+      internal app state. This is typically only done when dispatch is called
+      manually. Internal calls to dispatch are done via changeState which
+      updates the state itself so it's not needed to do it here.
   */
-  dispatch() {
+  dispatch(nullKeys = [], updateState = true) {
     let error, state;
     ({error, state} = this.generateState(this.location.href));
-    this._appStateHistory.push(state);
     if (error !== null) {
-      return `unable to generate state: ${error}`;
+      error += `unable to generate state: ${error}`;
+      return {error, state};
```

Comment:
It's possible that part of state was still generated, but an error was also reported. I like including whatever data  we have even when also returning an error.